### DISCUSSION TITLE
Implementing Challenge Solver in other language

### DISCUSSION THREAD

**Discussion by Dr-Blank** [tranxuanthang/lrcget#23] [Q&A] [answered]
I asked copilot to translate the challenge_solver.rs file and it came up with something. However, I am not sure how this code works; I understand that it needs to work out the nonce that was used given the prefix and the taget_hex by repeatedly hashing the "{prefix}{nonce}" string and comparing it to target.
It would be nice if you could provide some test cases that I can run to see if the implementation is correct as right now there is no way for me to verify.

**Comment by tranxuanthang** [comment, marked as answer]
> I understand that it needs to work out the nonce that was used given the prefix and the taget_hex by repeatedly hashing the "{prefix}{nonce}" string and comparing it to target.

You understood it correctly. In short, you must find an unknown `nonce` so that the the sha256 hash of `{prefix}{nonce}` is *smaller* than the `target`.

The current `target`'s hex:

```
000000FF00000000000000000000000000000000000000000000000000000000
```

Is present in an u8 array as:

```
[0, 0, 0, 255, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
```

For example, the following prefix:

```
kyhYEjsjkxcxHRliu3oOLQUodVumo6ur
```

Then the first `nonce` (from zero) that is valid is:

```
42660891
```

Because the sha256 of `kyhYEjsjkxcxHRliu3oOLQUodVumo6ur42660891` is:

```
000000C17F529D77D963888A368D9C13B5F0A9FC89E4BCBB5B213DF3AE4CA660
```

Or, in u8 array:

```
[0, 0, 0, 193, 127, 82, 157, 119, 217, 99, 136, 138, 54, 141, 156, 19, 181, 240, 169, 252, 137, 228, 188, 187, 91, 33, 61, 243, 174, 76, 166, 96]
```

Is smaller than the current `target`.

So you got the Publish Token with the value `kyhYEjsjkxcxHRliu3oOLQUodVumo6ur:42660891`

For high-level languages like Python, you may just hash it to a hex string and compare the 2 strings directly:

```
>>> '000000FF00000000000000000000000000000000000000000000000000000000' > '000000C17F529D77D963888A368D9C13B5F0A9FC89E4BCBB5B213DF3AE4CA660'
True
```

**Reply by Dr-Blank** [reply]
thank you for the explanation 🙌

**Reply by Dr-Blank** [reply]
I just have a minor doubt, does the nonce need to be the smallest one possible?

for your example, I found other nonce that was higher than `42660891` and satisfied the condition.

would that be rejected by the Api if a higher nonce is used or that is valid as well?

**Reply by tranxuanthang** [reply]
It doesn't have to. Any nonce that satisfies the condition will be accepted.

**Reply by Dr-Blank** [reply]
Could I ask the way prefix and target are generated or would that compromise the whole reason proof work is used in the first place?

I am curious because making sure target starts with zeros would require your server to do the work as well. And unless you have TPU at your disposal, how would Api return the challenge almost instantaneously. 

Unless you have already have done some work in advance and returning challenge on demand.

This has nothing to do with the question, just curiosity.

**Reply by Dr-Blank** [reply]
I just realized that the target string is the same for all challenges, that answers my question that the prefixes are random strings.

**Reply by tranxuanthang** [reply]
> making sure target starts with zeros would require your server to do the work as well

Target is predefined and rarely being changed. The client has to do the heavy work: brute force to find the valid `nonce` number. The server only needs to verify it against the predefined target.

> the target string is the same for all challenges

One note is that target might be changed in the future, so it's best that you use the target string in the `/api/request-challenge` response when solving challenge instead of hard coding it.

**Reply by Dr-Blank** [reply]
Thank you for all the help! I learnt something about cryptography while making a wrapper for this. I also learnt how slow python is 🥹. Your rust implementation takes seconds to compute the proof of work will my implementation takes a good 30 seconds even with multiple threads and processes.

**Reply by tranxuanthang** [reply]
You may implement the performance-critical part of your library using another low-level language, such as C, C++ or Rust. I cannot help you more as my Python knowledge is limited, but something like `pyo3` is a good starting point, or maybe `wasmer-python` for easy cross-platform support while still fast enough.

**Reply by Dr-Blank** [reply]
Yes, Having it in level language would help. Thank you for the guidance 🙌
